src/lib.rs

```diff
@@ -20,6 +20,77 @@ use std::{
 #[cfg(feature = "serde")]
 mod serde;
 
+/// Number type that's capable of representing [0, usize::MAX - 1]
+#[repr(transparent)]
+#[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd)]
+struct NonMaxUsize(NonZeroUsize);
+
+impl Debug for NonMaxUsize {
+  fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+    write!(f, "{}", self.get())
+  }
+}
+
+impl NonMaxUsize {
+  /// Convert an index to a usize
+  #[inline]
+  const fn get(&self) -> usize {
+    self.0.get() - 1
+  }
+
+  /// Create a new index from a usize, if `index` is `usize::MAX` then `None` is returned
+  #[inline]
+  const fn new(index: usize) -> Option<Self> {
+    match NonZeroUsize::new(index.wrapping_add(1)) {
+      Some(index) => Some(Self(index)),
+      None => None,
+    }
+  }
+
+  /// Create a new index from a usize, without checking if `index` is `usize::MAX`.
+  ///
+  /// # Safety
+  ///
+  /// `index` must not be `usize::MAX`
+  #[inline]
+  const unsafe fn new_unchecked(index: usize) -> Self {
+    Self(unsafe { NonZeroUsize::new_unchecked(index.wrapping_add(1)) })
```

Comment:
Is the wrapping_add needed here? It seems we can just do `i + 1` since it will be undefined behavior either way?

Reply:
Good point! I believe you are right.